File: Python/Sorting/Quick_Sort.py
def partition(A, low, high):
    pivot = A[low]
    i = low + 1
    j = high
    while True:
        while i <= j and A[i] <= pivot:
            i = i + 1
        while i <=j and A[j] > pivot:
            j = j - 1
        if i <= j:
            A[i], A[j] = A[j], A[i]
        else:
            break
    A[low], A[j] = A[j], A[low]
    return j


def quicksort(A, low, high):
    if low < high:
        pi = partition(A, low, high)
        quicksort(A, low, pi - 1)
        quicksort(A, pi + 1, high)



A = [6, 20, 815, 94, 651, 28]
print('Original Array:',A)
quicksort(A, 0, len(A)-1)
print('Sorted Array:',A)

File: Add Code Here/JAVA/Sorting/QuickSort.java
class QuickSort {
    // Function to partition the array into two subarrays
    int partition(int arr[], int low, int high) {
        int pivot = arr[high];
        int i = (low - 1); // Index of smaller element

        for (int j = low; j < high; j++) {
            // If the current element is smaller than or equal to the pivot
            if (arr[j] <= pivot) {
                i++;

                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Swap arr[i+1] and arr[high] (or the pivot)
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    // Function to perform Quick Sort
    void quickSort(int arr[], int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);

            // Recursively sort the elements before and after partition
            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    // Prints the array
    void printArray(int arr[]) {
        int n = arr.length;
        for (int i = 0; i < n; ++i) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static void main(String args[]) {
        QuickSort obj = new QuickSort();
        int arr[] = {64, 34, 25, 12, 9, 11, 90};
        int n = arr.length;
        System.out.println("Original array");
        obj.printArray(arr);
        obj.quickSort(arr, 0, n - 1);
        System.out.println("Sorted array");
        obj.printArray(arr);
    }
}


// Time Complexity: O(n log n)
// Space Complexity: O(log n) 